update/stock_update.js

```diff
@@ -1,122 +1,128 @@
-            var docs = data.rows.map(function(row) {
-                return row.doc;
-            });
-            callback(err, docs);
-        }
-    });
-}
-
-//include _id and _rev of existing doc in the doc to perform an update
-function insertOrUpdateDoc(doc, callback) {
-
-    db.insert(doc, function(err, body, header) {
-        if (err) {
-            console.log('insertion failed: ' + err.message);
-        } else if (isFunc(callback)) {
-            callback(body, header);
-        }
-    });
-}
-
 function findStockDatum(stocks, company) {
 
-    for (var i = 0; i < stocks.length; i++) {
-        var stock = stocks[i];
-        if (stock.ticker == company) {
-            return stock;
-        }
+  for (var i = 0; i < stocks.length; i++) {
+    var stock = stocks[i];
+    if (stock.ticker == company) {
+      return stock;
     }
+  }
 
-    return undefined;
+  return undefined;
 }
 
 function updateStocksData(articleData, stockData) {
+  
+  for (var i = 0; i < articleData.length; i++) {
+    var articleDatum = articleData[i];
+    var company = articleDatum.company
+    var articles = articleDatum.articles;
+    var stockDatum = findStockDatum(stockData, company);
+    if (stockDatum) {
+      stockDatum.history = articles.concat(stockDatum.history);
+    } else {
+      stockDatum = {
+        ticker : company,
+        history : articles
+      };
+    }
+    
+    //TODO batch insert?
+    console.log('Inserting into company: ' + company + ' articles: ' );
+    console.log(articles);
+    console.log();
+    stock_db.insertOrUpdateDoc(stockDatum);
+  }
+}
 
-    for (var i = 0; i < articleData.length; i++) {
-        var articleDatum = articleData[i];
-        var stockDatum = findStockDatum(stockData, articleDatum.company);
+function parseArticle(result) {
+  return {
+    url: result.url,
+    sentiment: result.enriched_text.sentiment.document.label,
+    date: result.crawl_date
+  }
+}
 
-        if (stockDatum) {
-            stockDatum.history.unshift(articleDatum.article);
-        } else {
-            stockDatum = {
-                ticker: articleDatum.company,
-                history: [articleDatum.article]
-            };
-        }
+function parseResults(results) {
+  var articles = [];
+  for (var i=0; i<results.length; i++) {
+    articles.push(parseArticle(results[i]));
+  }
+  return articles;
+}
 
-        //TODO batch insert?
-        console.log('Inserting into company ' + articleDatum.company + ' article: ' );
-        console.log(articleDatum.article);
-        insertOrUpdateDoc(stockDatum);
+function getArticleDataForCompany(company, callback) {
+  
+  var promise = discovery.query(company);
+    
+  promise.then(function (data) {
+    var results = data.results;
+    console.log("Received " + results.length + " articles for: " + company);
+    var articles = parseResults(results);
+    var data = {
+      company : company,
+      articles : articles
     }
+    callback(undefined, data);
```

Comment:
these callbacks look funny, why is this undefined? (note I'm a nodeJS noob)

Reply:
Only because of the ordering of the parameters. The callback has the error as the first arg, so we have to pass undefined so that data is the second arg. If we flipped the ordering, though, we could avoid this and pass (data, error) or (data) and both would work since error would be undefined anyway in the first case.